polkadot/node/network/collator-protocol/src/validator_side_experimental/peer_manager/connected.rs

```diff
@@ -0,0 +1,282 @@
+		Self { per_para, peer_info: Default::default() }
+	}
+
+	/// Update the reputation of a peer for a specific paraid. This needs to also be persisted to
+	/// the reputation database, but changes are duplicated to this in-memory store of connected
+	/// peers.
+	pub fn update_reputation(&mut self, update: ReputationUpdate) {
+		let Some(per_para) = self.per_para.get_mut(&update.para_id) else { return };
+		if u16::from(update.value) == 0 {
+			return
+		}
+		per_para.update_reputation(update);
+	}
+
+	/// Try accepting a new peer. The connection must have already been established, but here we
+	/// decide whether to keep it or not. We don't know for which paraid the peer will collate.
+	pub async fn try_accept<
+		RepQueryFn: Fn(PeerId, ParaId) -> QueryFut,
+		QueryFut: Future<Output = Score>,
+	>(
+		&mut self,
+		reputation_query_fn: RepQueryFn,
+		peer_id: PeerId,
+		peer_info: PeerInfo,
+	) -> TryAcceptOutcome {
+		if self.contains(&peer_id) {
+			return TryAcceptOutcome::Added
+		}
+
+		let mut outcome = TryAcceptOutcome::Rejected;
+
+		match peer_info.state() {
+			PeerState::Collating(para_id) => {
+				let past_reputation = reputation_query_fn(peer_id, *para_id).await;
+				if let Some(per_para) = self.per_para.get_mut(&para_id) {
+					let res = per_para.try_accept(peer_id, past_reputation);
+					outcome = outcome.combine(res);
+				}
+			},
+			PeerState::Connected =>
+				for (para_id, per_para) in self.per_para.iter_mut() {
+					let past_reputation = reputation_query_fn(peer_id, *para_id).await;
+					let res = per_para.try_accept(peer_id, past_reputation);
+					outcome = outcome.combine(res);
+				},
+		}
+
+		if !matches!(outcome, TryAcceptOutcome::Rejected) {
+			self.peer_info.insert(peer_id, peer_info);
+		}
+
+		outcome
+	}
+
+	/// Remove the peer. Should be called when we see a peer being disconnected.
+	pub fn remove(&mut self, peer: &PeerId) {
+		for per_para in self.per_para.values_mut() {
+			per_para.remove(peer);
+		}
+
+		self.peer_info.remove(peer);
+	}
+
+	/// Process a peer's declaration of intention to collate for this paraid.
+	pub fn declared(&mut self, peer_id: PeerId, para_id: ParaId) -> DeclarationOutcome {
+		let mut outcome = DeclarationOutcome::Rejected;
+
+		let Some(peer_info) = self.peer_info.get_mut(&peer_id) else { return outcome };
+
+		match peer_info.state() {
+			PeerState::Connected => {
+				for (para, per_para) in self.per_para.iter_mut() {
+					if para == &para_id && per_para.contains(&peer_id) {
+						outcome = DeclarationOutcome::Accepted;
+					} else {
+						// We remove the reserved slots from all other paras.
+						per_para.remove(&peer_id);
+					}
+				}
+			},
+			PeerState::Collating(old_para_id) if old_para_id == &para_id => {
+				// Redundant, already collating for this para.
+				outcome = DeclarationOutcome::Accepted;
+			},
+			PeerState::Collating(old_para_id) => {
+				if let Some(old_per_para) = self.per_para.get_mut(&old_para_id) {
+					old_per_para.remove(&peer_id);
```

Comment:
This needs to be changed. We want to support a shared collator infrastructure, so you can have same collator declaring on multiple paras. It is unlikely, but not impossible that in practice you can have the same collator be connected to same backer and advertise 2 collations for 2 diff paras.

Reply:
This is currently impossible (not only in this new implementation but in the existing one in production).
Collators send `Declare` messages which contain the paraid and then all further advertisements are assumed to be for that para. If they are not, they get punished via rep and the candidate is not backed.

So they can switch the paraid they are collating for but only sequentially (one at a time)

Reply:
Yeah, we need support this at some point, but I agree  not at this stage in the revamp.